### PR TITLE
Improve flow() typing

### PR DESCRIPTION
**Pull request by ismailhabib**
See: https://github.com/mobxjs/mobx/issues/1816

### CODE REVIEW

**Comment by ismailhabib** [comment]
Impact: before this MR, flow() only need a generic type when the passed function does not have any parameter. Now, it's always required (for a good reason).

File: src/api/flow.ts [hunk 1]
```diff
@@ -10,32 +10,9 @@ export interface FlowIterator<T> {
...
-export function flow<A1, A2, A3, A4, A5, A6, A7>(
-    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => FlowIterator<any>
-): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => CancellablePromise<any>
-export function flow<A1, A2, A3, A4, A5, A6, A7, A8>(
-    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => FlowIterator<any>
-): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => CancellablePromise<any>
-export function flow(generator: Function) {
+export function flow<T, U extends any[]>(
+    generator: (...args: U) => FlowIterator<any>
+): CancellablePromise<U> {
```
Comment (mayorovp) [inline src/api/flow.ts]: Why `<U>`?

Comment (ismailhabib) [reply src/api/flow.ts]: No significant meaning here. It can also be changed to <R, A> to follow the previous naming.

Comment (mayorovp) [reply src/api/flow.ts]: No, it cannot. Right result type is `(...args: U) => CancellablePromise<any>`

Comment (ismailhabib) [reply src/api/flow.ts]: You're right, I missed the `() =>` part. Strange that no tests are failing.

**Comment by xaviergonz** [comment]
There's a way to make flow properly detect the return argument as well (except when the return argument is a promise, but that can be circumvented with a fake function such as castFlowReturn or similar)

See https://github.com/mobxjs/mobx-state-tree/blob/5775437e5267281906ca2ef73a49e7e1bbb410bc/packages/mobx-state-tree/src/core/flow.ts for example

File: src/api/flow.ts [hunk 2]
```diff
@@ -10,32 +10,9 @@ export interface FlowIterator<T> {
...
-export function flow<A1, A2, A3, A4, A5, A6, A7>(
-    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => FlowIterator<any>
-): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => CancellablePromise<any>
-export function flow<A1, A2, A3, A4, A5, A6, A7, A8>(
-    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => FlowIterator<any>
-): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => CancellablePromise<any>
-export function flow(generator: Function) {
+export function flow<T, U extends any[]>(
+    generator: (...args: U) => FlowIterator<any>
+): (...args: U) => CancellablePromise<U> {
```
Comment (mayorovp) [inline src/api/flow.ts]: `<U>` again?

Comment (ismailhabib) [reply src/api/flow.ts]: fixed

**Pull request merged**